File: 백준/트리/2094(강수량).cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int getRangeMax(vector<int>& rainIndexTree, int left, int right) {
	left += rainIndexTree.size() / 2;
	right += rainIndexTree.size() / 2;
	int maxRain = 0;
	while (left <= right) {
		if (left % 2 == 1) {
			maxRain = max(maxRain, rainIndexTree[left]);
		}
		if (right % 2 == 0) {
			maxRain = max(maxRain, rainIndexTree[right]);
		}

		left = (left + 1) / 2;
		right = (right - 1) / 2;
	}
	return maxRain;
}
int getYearIdx(vector<int>& years, int key) {
	// 동일한 값이 있으면 해당 값의 index, 없으면 상한값의 index return

	int lo = -1;
	int hi = years.size() - 1;
	while (lo + 1 < hi) {
		int mid = lo + (hi - lo) / 2;
		if (years[mid] >= key) {
			hi = mid;
		}
		else {
			lo = mid;
		}
	}

	if (years[hi] < key) {
		return years.size();
	}
	return hi;
}
int solve(vector<int>& rainIndexTree, vector<int>& years, int y, int x) {
	
	// Y, X 존재하는지 확인
	int yIdx = getYearIdx(years, y);
	int xIdx = getYearIdx(years, x);
	
	bool existY = (yIdx < years.size() && years[yIdx] == y);
	bool existX = (xIdx < years.size() && years[xIdx] == x);

	
	// Y < Z < X를 만족하는 최대 강수량 구하기
	int rangeMax;
	int nextYIdx = getYearIdx(years, y + 1);
	if (nextYIdx >= years.size()) {
		rangeMax = 0;
	}
	else {
		rangeMax = getRangeMax(rainIndexTree, nextYIdx, xIdx - 1);
	}
	
	// 구한 정보들 이용해 결과 반환
	int startIdx = rainIndexTree.size() / 2;
	if (existY && existX && rainIndexTree[startIdx + xIdx] > rainIndexTree[startIdx + yIdx]) {
		return 1;
	}
	else if (existX && rangeMax >= rainIndexTree[startIdx + xIdx]) {
		return 1;
	}
	else if (existY && rangeMax >= rainIndexTree[startIdx + yIdx]) {
		return 1;
	}
	else if (existY && existX && xIdx - yIdx == x - y) {
		return 0;
	}
	else {
		return 2;
	}
}
void makeIndexTree(vector<int>& rainIndexTree, vector<int>& rains) {
	int startIdx = 1;
	while (startIdx < rains.size()) {
		startIdx *= 2;
	}
	rainIndexTree.assign(startIdx * 2, 0);
	for (int i = 0; i < rains.size(); ++i) {
		rainIndexTree[i + startIdx] = rains[i];
	}
	for (int i = startIdx - 1; i > 0; --i) {
		rainIndexTree[i] = max(rainIndexTree[i * 2], rainIndexTree[i * 2 + 1]);
	}
}
void input(int& n, vector<int>& years, vector<int>& rains) {
	cin >> n;
	years.assign(n, 0);
	rains.assign(n, 0);
	for (int i = 0; i < n; ++i) {
		cin >> years[i] >> rains[i];
	}
}
int main() {
	int n;
	vector<int> years;
	vector<int> rains;
	input(n, years, rains);

	vector<int> rainIndexTree;
	makeIndexTree(rainIndexTree, rains);

	int m;
	cin >> m;
	vector<string> answers(m, "");
	for (int i = 0; i < m; ++i) {
		int y, x;
		cin >> y >> x;
		int state = solve(rainIndexTree, years, y, x);

		if (state == 0) {
			answers[i] = "true";
		}
		else if (state == 1) {
			answers[i] = "false";
		}
		else {
			answers[i] = "maybe";
		}
	}

	for (int i = 0; i < m; ++i) {
		cout << answers[i] << '\n';
	}
	return 0;
}

File: 백준/트리/2094(강수량).java
import java.util.*;
import java.io.*;

public class Main {

    private static int n, m;
    private static int[] years;
    private static int[] rains;

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        input(br);

        StringBuilder sb = new StringBuilder();
        solve(br, sb);

        System.out.print(sb);
        br.close();
    }

    private static void input(BufferedReader br) throws IOException {

        n = Integer.parseInt(br.readLine());
        years = new int[n];
        rains = new int[n];

        StringTokenizer st = null;
        for(int i = 0; i < n; ++i) {
            st = new StringTokenizer(br.readLine());

            years[i] = Integer.parseInt(st.nextToken());
            rains[i] = Integer.parseInt(st.nextToken());
        }
    }

    private static void solve(BufferedReader br, StringBuilder sb) throws IOException {
        int[] maxIDT = makeMaxIDT();

        m = Integer.parseInt(br.readLine());

        StringTokenizer st = null;
        int Y, X;
        int idxY, idxX, nextY;
        boolean existY, existX;
        int maxRain;
        for(int i = 0; i < m; ++i) {
            st = new StringTokenizer(br.readLine());
            Y = Integer.parseInt(st.nextToken());
            X = Integer.parseInt(st.nextToken());

            idxY = binarySearch(Y);
            idxX = binarySearch(X);

            existY = (idxY < n && years[idxY] == Y);
            existX = (idxX < n && years[idxX] == X);

            nextY = binarySearch(Y + 1);
            if(nextY == n) {
                maxRain = 0;
            } else {
                maxRain = getRangeMax(maxIDT, nextY, idxX - 1);
            }

            getAnswer(sb, existY, existX, idxY, idxX, maxRain);
        }
    }

    private static int[] makeMaxIDT() {
        //IDT 크기 설정
        int startIdx = 1;
        while(startIdx < n) {
            startIdx *= 2;
        }

        int[] maxIDT = new int[startIdx * 2];

        //leaf node 채우기
        for(int i = 0; i < n; ++i) {
            maxIDT[i + startIdx] = rains[i];
        }

        //내부 노드 채우기
        for(int i = startIdx - 1; i >= 1; --i) {
            maxIDT[i] = Math.max(maxIDT[i * 2], maxIDT[i * 2 + 1]);
        }

        return maxIDT;
    }

    private static int binarySearch(int key) {
        int lo = -1;
        int hi = n - 1;
        while(lo + 1 < hi) {
            int mid = lo + (hi - lo) / 2;
            if(years[mid] < key) {
                lo = mid;
            } else {
                hi = mid;
            }
        }

        if(years[hi] < key) {
            return n;
        }
        return hi;
    }

    private static int getRangeMax(int[] maxIDT, int left, int right) {
        left += (maxIDT.length / 2);
        right += (maxIDT.length / 2);
        int max = 0;
        while(left <= right) {
            if(left % 2 == 1) {
                max = Math.max(max, maxIDT[left]);
            }
            if(right % 2 == 0) {
                max = Math.max(max, maxIDT[right]);
            }

            left = (left + 1) / 2;
            right = (right - 1) / 2;
        }
        return max;
    }

    private static void getAnswer(StringBuilder sb, boolean existY, boolean existX, int idxY, int idxX, int maxRain) {

        if(existY && existX) {
            if(rains[idxX] > rains[idxY]) {
                sb.append("false\n");
            } else {
                if(maxRain >= rains[idxX]) {
                    sb.append("false\n");
                } else {
                    if((years[idxX] - years[idxY]) == (idxX - idxY)) {
                        sb.append("true\n");
                    } else {
                        sb.append("maybe\n");
                    }
                }
            }
        } else if(existY) {
            if(maxRain >= rains[idxY]) {
                sb.append("false\n");
            } else {
                sb.append("maybe\n");
            }
        } else if(existX) {
            if(maxRain >= rains[idxX]) {
                sb.append("false\n");
            } else {
                sb.append("maybe\n");
            }
        } else {
            sb.append("maybe\n");
        }
    }
    
}

